File: Components/Strata.Outbox.Client/src/main/typescript/RequestReply/OutboxServiceClient.ts
import {IOutboxServiceClient} from "./IOutboxServiceClient";
import {ICompletionStage} from "strata.foundation.core/Concurrent";
import {AbstractRestClient} from "strata.client.core/Service";
import {Guid} from "guid-typescript";
import {AbstractServiceRequest} from "strata.foundation.core/Transfer";
import {Instant} from "@js-joda/core";
import {
    CreateOutboxReply,
    CreateOutboxRequest,DestroyOutboxReply,
    DestroyOutboxRequest,FindOutboxReply,FindOutboxRequest,
    UpdateOutboxReply,
    UpdateOutboxRequest
} from "strata.outbox.service/RequestReply";

export
class OutboxServiceClient
    extends AbstractRestClient
    implements IOutboxServiceClient
{
    constructor(baseUrl: string)
    {
        super(baseUrl);
        this.setHeader("Content-Type","application/json");
        this.setHeader("Access-Control-Allow-Origin","*");
    }

    setAuthorization(accessToken: string): IOutboxServiceClient
    {
        if (super.hasHeader("Authorization"))
            super.clearHeader("Authorization");

        super.setHeader("Authorization","Bearer " + accessToken);
        return this;
    }

    close(): void
    {
        super.clearHeaders();
    }

    createOutbox(request: CreateOutboxRequest): ICompletionStage<CreateOutboxReply>
    {
        console.log("OutboxServiceClient.createOutbox");
        return this.doPost<CreateOutboxReply,CreateOutboxRequest>(
            "create-outbox",
            this.initializeRequest(request));
    }

    updateOutbox(request: UpdateOutboxRequest): ICompletionStage<UpdateOutboxReply>
    {
        console.log("OutboxServiceClient.updateOutbox");
        return this.doPost<UpdateOutboxReply,UpdateOutboxRequest>(
            "update-outbox",
            this.initializeRequest(request));
    }

    destroyOutbox(request: DestroyOutboxRequest): ICompletionStage<DestroyOutboxReply>
    {
        console.log("OutboxServiceClient.destroyOutbox");
        return this.doPost<DestroyOutboxReply,DestroyOutboxRequest>(
            "destroy-outbox",
            this.initializeRequest(request));
    }

    findOutbox(request: FindOutboxRequest): ICompletionStage<FindOutboxReply>
    {
        console.log("OutboxServiceClient.find-outbox");
        return this.doPost<FindOutboxReply,FindOutboxRequest>(
            "find-outbox",
            this.initializeRequest(request));
    }

    private  initializeRequest<R extends AbstractServiceRequest>(request: R): R
    {
        console.log("OutboxServiceClient.initializeRequest");
        if (request == null)
            request.requestId = Guid.create().toString();

        if (request.timestamp == null)
            request.timestamp = Instant.now().toString();

        return request;
    }

}

File: Components/Strata.Outbox.Client/src/main/java/strata/outbox/client/requestreply/OutboxServiceClient.java
//////////////////////////////////////////////////////////////////////////////
// OutboxServiceClient.java
//////////////////////////////////////////////////////////////////////////////

package strata.outbox.client.requestreply;

import strata.outbox.service.requestreply.*;
import jakarta.ws.rs.client.ClientBuilder;
import org.apache.hc.core5.ssl.SSLContextBuilder;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.Resource;
import strata.client.core.service.AbstractRestClient;

import javax.net.ssl.SSLContext;
import java.util.concurrent.CompletionStage;

public
class OutboxServiceClient
    extends AbstractRestClient
    implements IOutboxServiceClient
{
    public
    OutboxServiceClient(String baseUrl)
    {
        super(
            createBuilder(),
            preprocessBaseUrl(baseUrl),
            "outbox-service/");
    }

    @Override
    public CompletionStage<StartWorkerReply>
    startWorker(StartWorkerRequest request)
    {
        return
            doPostAsync(
                "create-outbox",
                StartWorkerReply.class,
                request);
    }

    @Override
    public CompletionStage<StopWorkerReply>
    stopWorker(StopWorkerRequest request)
    {
        return
            doPostAsync(
                "update-outbox",
                StopWorkerReply.class,
                request);
    }

    @Override
    public CompletionStage<QueryWorkerReply>
    queryWorker(QueryWorkerRequest request)
    {
        return
            doPostAsync(
                "destroy-outbox",
                QueryWorkerReply.class,
                request);
    }

    @Override
    public CompletionStage<FindOutboxReply>
    findOutbox(FindOutboxRequest request)
    {
        return
            doPostAsync(
                "find-outbox",
                FindOutboxReply.class,
                request);
    }

    public OutboxServiceClient
    setAuthorizationHeader(String accessToken)
    {
        setHeader("Authorization","Bearer " + accessToken);
        return this;
    }

    private static String
    preprocessBaseUrl(String baseUrl)
    {
        return
            baseUrl.endsWith("/")
                ? baseUrl
                : baseUrl + '/';
    }

    private static ClientBuilder
    createBuilder() throws RuntimeException
    {
        return
            ClientBuilder
                .newBuilder()
                .sslContext(createSslContext());
    }

    private static SSLContext
    createSslContext() throws RuntimeException
    {
        try
        {
            Resource keystore = new ClassPathResource("hello-service.p12");
            return
                new SSLContextBuilder()
                    .loadTrustMaterial(
                        keystore.getURL(),
                        "hello-service".toCharArray())
                    .build();
        }
        catch (Exception e)
        {
            throw new RuntimeException(e);
        }

    }
}

//////////////////////////////////////////////////////////////////////////////
